netty/src/main/java/io/grpc/netty/GrpcHttp2HeadersDecoder.java

```diff
@@ -0,0 +1,499 @@
+  }
+
+  private final class GrpcHttp2HeaderTable implements Http2HeaderTable {
+
+    private int maxHeaderListSize = Integer.MAX_VALUE;
+
+    @Override
+    public void maxHeaderTableSize(int max) throws Http2Exception {
+      if (max < 0) {
+        throw connectionError(PROTOCOL_ERROR, "Header Table Size must be non-negative but was %d",
+            max);
+      }
+      try {
+        decoder.setMaxHeaderTableSize(max);
+      } catch (Throwable t) {
+        throw connectionError(PROTOCOL_ERROR, t.getMessage(), t);
+      }
+    }
+
+    @Override
+    public int maxHeaderTableSize() {
+      return decoder.getMaxHeaderTableSize();
+    }
+
+    @Override
+    public void maxHeaderListSize(int max) throws Http2Exception {
+      if (max < 0) {
+        // Over 2^31 - 1 (minus in integer) size is set to the maximun value
+        maxHeaderListSize = Integer.MAX_VALUE;
+      } else {
+        maxHeaderListSize = max;
+      }
+    }
+
+    @Override
+    public int maxHeaderListSize() {
+      return maxHeaderListSize;
+    }
+  }
+
+  static final class GrpcHttp2ServerHeadersDecoder extends GrpcHttp2HeadersDecoder {
+
+    GrpcHttp2ServerHeadersDecoder(int maxHeaderSize) {
+      super(maxHeaderSize);
+    }
+
+    @Override GrpcHttp2InboundHeaders newHeaders(int numHeadersGuess) {
+      return new GrpcHttp2RequestHeaders(numHeadersGuess);
+    }
+  }
+
+  static final class GrpcHttp2ClientHeadersDecoder extends GrpcHttp2HeadersDecoder {
+
+    GrpcHttp2ClientHeadersDecoder(int maxHeaderSize) {
+      super(maxHeaderSize);
+    }
+
+    @Override GrpcHttp2InboundHeaders newHeaders(int numHeadersGuess) {
+      return new GrpcHttp2ResponseHeaders(numHeadersGuess);
+    }
+  }
+
+  /**
+   * A {@link Http2Headers} implementation optimized for inbound/received headers.
+   *
+   * <p>Header names and values are stored in simple arrays, which makes insert run in O(1)
+   * and retrievial a O(n). Header name equality is not determined by the equals implementation of
+   * {@link CharSequence} type, but by comparing two names byte to byte.
+   *
+   * <p>All {@link CharSequence} input parameters and return values are required to be of type
+   * {@link AsciiString}.
+   */
+  abstract static class GrpcHttp2InboundHeaders extends AbstractHttp2Headers {
+
+    private byte[][] namesAndValues;
+    private AsciiString[] values;
+    private int namesAndValuesIdx;
+
+    GrpcHttp2InboundHeaders(int numHeadersGuess) {
+      checkArgument(numHeadersGuess > 0, "numHeadersGuess needs to be gt zero.");
+      namesAndValues = new byte[numHeadersGuess << 1][];
+      values = new AsciiString[numHeadersGuess];
+    }
+
+    protected Http2Headers add(AsciiString name, AsciiString value) {
+      if (namesAndValuesIdx == namesAndValues.length) {
+        expandHeadersAndValues();
+      }
+      byte[] nameBytes = bytes(name);
+      byte[] valueBytes = toRawSeralizedHeader(nameBytes, bytes(value));
+      values[namesAndValuesIdx >> 1] = value;
```

Comment:
Oh me too :-). Then I misunderstood. Cause I had it that way in the first place. They are the same speed, cause static powers of two get compiled to shift instructions. I ll change them back :)